### PR TITLE
Support for Fixed Sized Arrays

### PR DESCRIPTION
**Pull request by Carter12s**
## Description
ROS1 serialization with serde_rosmsg requires that fixed sized arrays be represented not as Vec<_> but as [_; N]. The two are sent over the wire differently. Most of the support for this was already plumbed through, so this just closed the loop at the end, but is hitting the "const generic" barrier.

## Fixes
Closes: #148 

## Checklist
- [x] Update CHANGELOG.md

### CODE REVIEW

**Comment by Carter12s** [comment]
@ssnover Do you have any good ideas on how we should handle this situtation?

1. Is there a workaround where maybe we can use some other type than [T; N] to represent a fixed sized array?
2. Can we codegen our own trait impls for larger array sizes? (This seems awful)
3. Do we just declare our crate only works with fixed sized arrays <32 elements until const generics land? can we limit that impact to ROS1 usage only? probably not cause it affects codegen...

Am I just missing an obvious solve here.

**Comment by ssnover** [comment]
We have at least one instance showing in the codegen for a 36 member array, so we probably ought to at least support that as a basic ROS message. 

Only alternative I can think of is to represent with a tuple of Vec and the fixed size and then define our own serialization and deserialization for serde_rosmsg?

**Comment by Carter12s** [comment]
I have a little side test setup in my fork of serde_rosmsg, from when the bug was originally reported, where I'm deserializing a NavSatFix.msg with the library. NavSatFix has an [f64; 9] in it.

For Box<[f64]> serde_rosmsg spat out an underflow error at runtime.
For &[f64] we get a compile time "Deserialize<'_> is not implemented"

So don't see a good path there...
I may open an issue on serde_rosmsg and see if they have a solve...

**Comment by Carter12s** [comment]
https://stackoverflow.com/questions/62665558/how-can-i-implement-serdedeserialize-for-arrays-larger-than-32

Has lead me to: https://crates.io/crates/serde-big-array

Which I'm going to play with and see if it can solve this problem...

**Comment by Carter12s** [comment]
So it looks like a combination of some fuckery with smart_default's _code directive and this serde-big-array crate I may have gotten a working solution.

Biggest problem I see is 1) the code is super messy and could use some unit tests, and 2) we're leaking yet another crate dependency through the codegen interface in not a great way...

I'd love your eyes on this @ssnover

File: roslibrust_codegen/src/gen.rs [hunk 1]
```diff
@@ -145,20 +147,48 @@ fn generate_field_definition(
...
+                let default_str = format!("[Default::default(); {fixed_array_length}]");
+                quote! { #[default(_code = #default_str)]}
+            } else {
+                quote! {}
+            }
+        } else {
+            quote! {}
+        }
+    };
+    let serde_line = if let Some(Some(fixed_array_length)) = field.field_type.array_info {
```
Comment (ssnover) [inline roslibrust_codegen/src/gen.rs:176]: Nitpick: This is a little unclean looking, perhaps:
```rust
let serde_line = match field.field_type.array_info {
    Some(Some(fixed_array_len)) if fixed_array_len > MAX_FIXED_ARRAY_LEN => quote! { #[serde(with = "::serde_big_array::BigArray")] },
    _ => quote! {},
};
```

File: roslibrust_codegen/src/parse/mod.rs [hunk 2]
```diff
@@ -216,3 +216,24 @@ fn parse_type(type_str: &str, pkg: &Package) -> Result<FieldType, Error> {
         }
     }
 }
+
+#[cfg(test)]
+mod test {
```
Comment (ssnover) [inline roslibrust_codegen/src/parse/mod.rs:221]: Big fan of introducing more test cases like this.

File: roslibrust_codegen/src/lib.rs [hunk 3]
```diff
@@ -631,7 +631,7 @@ fn parse_ros_files(
 
 #[cfg(test)]
 mod test {
-    use crate::find_and_generate_ros_messages;
+    use crate::{find_and_generate_ros_messages, FieldInfo};
```
Comment (ssnover) [inline roslibrust_codegen/src/lib.rs:634]: Looks like this was unused.

File: roslibrust/Cargo.toml [hunk 4]
```diff
@@ -50,6 +50,7 @@ regex = { version = "1.9", optional = true } # Only used with native ros1
 env_logger = "0.10"
 test-log = "0.2"
 simple_logger = "2.1.0"
+serde-big-array = "0.5"
```
Comment (ssnover) [inline roslibrust/Cargo.toml:53]: This was not picked up in the noetic CI. This is the only blocker I think.

**Review by ssnover — CHANGES_REQUESTED** [review]
This is great, thanks for taking the time to figure out the attribute macro magic required to get this functioning and adding tests.

I pointed out the current CI failures in comments.

File: roslibrust_codegen/src/gen.rs (hunk 1 above)
Comment (Carter12s) [reply roslibrust_codegen/src/gen.rs:176]: Love that much cleaner.

**Pull request merged**